core/src/main/java/org/elasticsearch/index/mapper/Uid.java

```diff
@@ -123,4 +126,169 @@ public static String createUid(String type, String id) {
+            if (c < '0' || c > '9') {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /** With numeric ids, we just fold two consecutive chars in a single byte
+     *  and use 0x0f as an end marker. */
+    private static BytesRef encodeNumericId(String id) {
+        byte[] b = new byte[1 + (id.length() + 1) / 2];
+        b[0] = (byte) NUMERIC;
+        for (int i = 0; i < id.length(); i += 2) {
+            int b1 = id.charAt(i) - '0';
+            int b2;
+            if (i + 1 == id.length()) {
+                b2 = 0x0f; // end marker
+            } else {
+                b2 = id.charAt(i + 1) - '0';
+            }
+            b[1 + i/2] = (byte) ((b1 << 4) | b2);
+        }
+        return new BytesRef(b);
+    }
+
+    /** With base64 ids, we decode and prepend an escape char in the cases that
+     *  it could be mixed up with numeric or utf8 encoding. In the majority of
+     *  cases (253/256) the encoded id is exactly the binary form. */
+    private static BytesRef encodeBase64Id(String id) {
+        byte[] b = Base64.getUrlDecoder().decode(id);
+        if (Byte.toUnsignedInt(b[0]) >= BASE64_ESCAPE) {
+            byte[] newB = new byte[b.length + 1];
+            newB[0] = (byte) BASE64_ESCAPE;
+            System.arraycopy(b, 0, newB, 1, b.length);
+            b = newB;
+        }
+        return new BytesRef(b, 0, b.length);
+    }
+
+    private static BytesRef encodeUtf8Id(String id) {
+        byte[] b = new byte[1 + UnicodeUtil.maxUTF8Length(id.length())];
+        // Prepend a byte that indicates that the content is an utf8 string
+        b[0] = (byte) UTF8;
+        int length = UnicodeUtil.UTF16toUTF8(id, 0, id.length(), b, 1);
+        return new BytesRef(b, 0, length);
+    }
+
+    /** Encode an id for storage in the index. This encoding is optimized for
+     *  numeric and base64 ids, which are encoded in a much denser way than
+     *  what UTF8 would do.
+     *  @see #decodeId */
+    public static BytesRef encodeId(String id) {
+        if (id.isEmpty()) {
+            throw new IllegalArgumentException("Ids can't be empty");
+        }
+        if (isPositiveNumeric(id)) {
+            // common for ids that come from databases with auto-increments
+            return encodeNumericId(id);
+        } else if (isURLBase64WithoutPadding(id)) {
+            // common since it applies to autogenerated ids
+            return encodeBase64Id(id);
+        } else {
+            return encodeUtf8Id(id);
+        }
+    }
+
+    private static String decodeNumericId(byte[] idBytes) {
+        assert Byte.toUnsignedInt(idBytes[0]) == NUMERIC;
+        int length = (idBytes.length - 1) * 2;
+        char[] chars = new char[length];
+        for (int i = 1; i < idBytes.length; ++i) {
+            final int b = Byte.toUnsignedInt(idBytes[i]);
+            final int b1 = (b >>> 4);
+            final int b2 = b & 0x0f;
+            chars[(i - 1) * 2] = (char) (b1 + '0');
```

Comment:
Maybe try using two 256 entry lookup tables for the upper and lower values. That would save two additions, the right shift, and cast from byte to int.